crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -233,11 +240,37 @@ public(package) fun request_add_validator(self: &mut ValidatorSet, min_joining_s
         EDuplicateValidator
     );
     assert!(validator.is_preactive(), EValidatorNotCandidate);
-    assert!(validator.total_stake_amount() >= min_joining_stake_amount, EMinJoiningStakeNotReached);
+    assert!(self.can_join(validator.total_stake_amount(), ctx), EMinJoiningStakeNotReached);
 
     self.pending_active_validators.push_back(validator);
 }
 
+/// Return `true` if a  candidate validator with `stake` will have sufficeint voting power to join the validator set
+fun can_join(self: &ValidatorSet, stake: u64, ctx: &TxContext): bool {
+    let (min_joining_voting_power, _, _) = self.get_voting_power_thresholds(ctx);
+
+    // if the validator will have at least `min_joining_voting_power` after joining, they can join.
+    // this formula comes from SIP-39 TODO link once landed
+    let future_total_stake = self.total_stake + stake;
+    let future_validator_voting_power = voting_power::derive_raw_voting_power(stake, future_total_stake);
+    future_validator_voting_power >= min_joining_voting_power
+}
+
+/// return (min, low, very low voting power) thresholds
+fun get_voting_power_thresholds(self: &ValidatorSet, ctx: &TxContext): (u64, u64, u64) {
+    let start_epoch = {
+        let key = VotingPowerAdmissionStartEpochKey();
+        if (self.extra_fields.contains(key)) self.extra_fields[key]
+        else ctx.epoch() + 1 // will give us the phase 1 values
+    };
+
+    // these numbers come from SIP-39 TODO link once landed
```

Comment:
you can add the link now https://github.com/sui-foundation/sips/blob/main/sips/sip-39.md